### PR TITLE
Add a simple amp-like jump command

### LINKED ISSUE #274: Better `f` à la clever-f and vim-sneak like jump motion
Since we're already deviating from the behavior of `f` in vim (multiline, not restricted to the current line), it'd we useful if we could go a step further and implement something like [clever-f](https://github.com/rhysd/clever-f.vim). It would enable pressing `f` again immediately after a previous `f` search (pressing any other key resets this) in lieu of `;`. This is really useful when you get the count wrong for `f` and want to spam `f` to go to the desired match.

Another really useful feature would be something along the lines of [vim-easymotion](https://github.com/easymotion/vim-easymotion) and [vim-sneak](https://github.com/justinmk/vim-sneak) for jumping anywhere in the view (uses two letters to find the jump point).

### LINKED ISSUE #510: AMP like Jump Mode
Amp Editor (https://amp.rs/docs/usage/) has a very sweet jump mode.
Since it's also written in Rust and open source, would it be a crime to suggest to peek into the code and steal some ideas from it?

### PR DESCRIPTION
**Pull request by pascalkuthe**
Closes #274
Closes #510
Closes #5340

This is a simple amp-like jumpmode that @the-mikedavis and I have been toying around with for a while. When discussing/starting to review #5340 we found the PR more complex than we would like for this feature (a diff above 1k loc). We were uncomfortable with how closely it follows hop and the subsequent debate about emulating the million other jumpmode plugins out there. There are many different approaches to jump commands, and we cannot maintain them all in core. Instead, we want to pick a single opinionated choice that we feel fits well with the rest of the editor.

While testing that PR we made the following observations:

* character based jumps (which require a character before tying a label) don't really hold their water, you can use `f` for simple cases and `/` (which we want to improve further) for complex cases. Compared to a jumpmode these commands work with multicursors.
* word base jumps can be useful for quickly repositioning the primary cursor.
* a simple implementation like in AMP without dimming, color changes and complex letter assignment based on distance, ... was not only a lot simpler to implement but also more predictable/less disruptive

Based on these observations we decided to mostly follow amps implementation. This PR also has a more correct implementation fixing issues I found with the original PR (labels would appear on single char words, punction and would include leading whitespace, ..).

This PR contains a cherry pick from #6417, depending on which PR is merged first I will rebase the other PR and drop that commit

### CODE REVIEW

File: book/src/keymap.md
Comment (the-mikedavis) [inline book/src/keymap.md:227]:
```suggestion
| `w`   | Show labels at each word and select the word that belongs to the entered labels | `goto_word` |

#### Match mode
```

this probably isn't an issue while rendering the markdown but it looks like it's breaking github's markdown syntax highlighting a little

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5767]:
```suggestion
fn jump_to_label(cx: &mut Context, labels: Vec<Range>, extend: bool) {
```

what do you think? I'm thinking that just "jump_to" is too generic

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5857]:
```suggestion
    let (view, doc) = current_ref!(cx.editor);
```

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5768]:
```suggestion
    let doc = doc!(cx.editor);
```

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5764]: We could re-use the `Movement` enum here instead of the bool

File: helix-term/src/commands.rs
Comment (matoous) [inline helix-term/src/commands.rs:5908]: **thought**: this could reduce the nesting and (subjectively) make it easier to read:
```suggestion
    let mut cursor_rev = cursor_fwd;

    // collect words after the cursor
    while cursor_fwd.head < end && words.len() < jump_label_limit {
        cursor_fwd = movement::move_next_word_end(text, cursor_fwd, 1);
        let range = movement::move_prev_word_start(text, cursor_fwd, 1);
        // The cursor is on a word longer than 2 characters.
        if RopeGraphemes::new(text.slice(range.head..))
            .take(2)
            .all(|g| g.chars().all(char_is_word))
        {
            words.push(range.flip());
        }
    }

    // collect words before the cursor
    while cursor_rev.head > start && words.len() < jump_label_limit {
        cursor_rev = movement::move_prev_word_start(text, cursor_rev, 1);
        // The cursor is on a word longer than 2 characters.
        if RopeGraphemes::new(text.slice(cursor_rev.head..))
            .take(2)
            .all(|g| g.chars().all(char_is_word))
        {
            words.push(movement::move_next_word_end(text, cursor_rev, 1).flip());
        }
    }
```
I hope I didn't miss something in regards to the logic.

Comment (pascalkuthe) [reply helix-term/src/commands.rs:5908]: that is not equivalent tough. I purposefully iterate forward and reverse in lockstep so that labels are assigned in order of the textual distance.

Comment (matoous) [reply helix-term/src/commands.rs:5908]: Ah, gotcha. I see now. Thank you.

File: helix-term/src/commands.rs
Comment (pascalkuthe) [reply helix-term/src/commands.rs]: yeah that seems reasonable , originally `jump_to_word` was actualy named `jump_to_label` so I couldn't call it that originally but I renamed that function after so now that name makes sense

File: book/src/keymap.md
Comment (JeftavanderHorst) [inline book/src/keymap.md:226]: The command is documented as `goto_word` but is actually called `goto_label`

Comment (pascalkuthe) [reply book/src/keymap.md:226]: its actually the documentation that is right here, I renamed the command in the codebase to match

**Comment by danillos** [comment]
I was testing it and I didn't saw any issue so far on the current implementation. Nice work.

I have a question, would it be possible to work in all visible buffers? So we could go to a word in another split window?

**Comment by David-Else** [comment]
I have been testing it and it works brilliantly, thanks!

**Comment by Omnikar** [comment]
Should the default value for `jump-label-alphabet` be reordered to prefer QWERTY home-row characters first?

**Review by archseer — APPROVED** [review]
Looks fine to me but I'm not a jump mode user. I'd be interested in @semin-park's feedback as the original PR owner

**Comment by David-Else** [comment]
> > Should the default value for `jump-label-alphabet` be reordered to prefer QWERTY home-row characters first?
> 
> We generally don't try to tailor our keymap to qwerty (many people use different keymaps). Using the alphabet is a pretty reasonal default IMO.
> 
> I have set the setting to my (colmak) homerow plus some keys that are easily reachabke. What exactly those are is highly individual.

Alphabet is certainly a good default, but having a config option to prefer the home row keys of QUERTY / Colmeak / Dvorak  would be next level awesome :)

> I have set the setting to my (colmak) homerow plus some keys

How is this possible? Cheers!

**Comment by pascalkuthe** [comment]
just add this to your config (under editor, its also documented in this PR):

jump-label-alphabet = "eirsaotndhfuywgm(cplbjvkx)z_"

(that example is for my personal layout where brackets and underscore are on the base layer)

**Comment by David-Else** [comment]
@pascalkuthe You are amazing, you thought of everything. Sorry I didn't see that in the PR and wasted your time.

**Comment by Omnikar** [comment]
> > Should the default value for `jump-label-alphabet` be reordered to prefer QWERTY home-row characters first?
> 
> We generally don't try to tailor our keymap to qwerty (many people use different keymaps).

Doesn't having HJKL as default directional controls already do that?

**Comment by Omnikar** [comment]
@qwerty01 I think the recommendation for that is to just use search (`/` and `?`)

**Comment by qwerty01** [comment]
`/` and `f` would work well for spots that are easily searchable, but braces and other non-word characters are usually not very searchable since they're 1 character and would have lots of matches. With the current solution, it would probably be finding some other spot on the line and then doing some other key combinations to get to the spot you want to, but that kinda defeats the purpose of amp jumps, which is looking at the spot you want to go and typing the keys that pop up.

**Comment by Omnikar** [comment]
Well you could also type characters before and/or after the brace you're trying to jump to when using `/`.

**Comment by qwerty01** [comment]
Oh quick aside, found an issue when scrolling while jump mode is active, only the visible screen at the time of enabling jump mode gives jump options, so if you scroll while jump mode is active, there aren't more options to jump:

scrolling

**Comment by the-mikedavis** [comment]
Yeah, I think we should clear the jump labels / exit jumping mode when the viewport changes (scrolling, resizing, etc.). It's only meant to jump to locations that are in-view.

> I have a question, would it be possible to work in all visible buffers? So we could go to a word in another split window?

Currently it's a regular motion so like `w`/`b`/`e` or text-objects it only works on the current view (window). IMO jumping across views is not necessary to optimize for - I think switching views and then jumping within the view would work well enough.

**Comment by pascalkuthe** [comment]
> Yeah, I think we should clear the jump labels / exit jumping mode when the viewport changes (scrolling, resizing, etc.). It's only meant to jump to locations that are visible/in-view.
> 
> > I have a question, would it be possible to work in all visible buffers? So we could go to a word in another split window?
> 
> Currently it's a regular motion so like `w`/`b`/`e` or text-objects it only works on the current view (window). IMO jumping across views is not necessary to optimize for - I think switching views and then jumping within the view would work well enough.

yeah I fixed this by simply dismissing any pending (or pseudo pending) keys whenever a mouse button is pressed (also during bracketed paste). I think mouse buttons behaving the same as other keys in that regard just makes sense. I think this will fix quite a few similar bugs.

**Comment by archseer** [comment]
> We generally don't try to tailor our keymap to qwerty (many people use different keymaps).

I don't have strong opinions here, but I wouldn't mind ordering the default to use the qwerty homerow. I also don't use qwerty but the majority of the users do and it would make the default config slightly more efficient for them (the rest of us will remap anyway).

File: helix-view/src/editor.rs
Comment (archseer) [inline helix-view/src/editor.rs:304]: We probably want to validate that the chars are unique with no duplicates

**Comment by postsolar** [comment]
I think a feature worth considering is jumping while adding the target word as a new selection, similarly to what `n` in select mode does.

**Comment by the-mikedavis** [comment]
We might want to consider that as a larger effort of https://github.com/helix-editor/helix/issues/5672. It might be useful to add selections rather than extend in some other cases like textobjects and tree-sitter motions. We could have a separate mode for it so that select mode always extends and the new mode always adds.

In any case though that should be left for future work rather than added to this PR

**Comment by brunoherrick** [comment]
This seems to implement a "hop.nvim" jump mechanism (2 random chars). Will this also include a "leap.nvim"/ "lightspeed.nvim" jump mechanism (2 chars + eventual 3rd)? Gauging from github and forums such as reddit, the "leap" strategy seems more popular and is implemented in the previous PR #5340 . If this is intended to be part of the core, I would suggest to also include the seemingly more popular approach.

**Comment by the-mikedavis** [comment]
Could you elaborate on the behavior you're looking for? I'm not sure anything like that is implemented in #5340. The character based jumps like with `g/` from #5340 I think would be ok to add as a follow-up to this PR as long as it only operates on the current view and re-uses most of the code introduced here.

**Comment by brunoherrick** [comment]
@the-mikedavis exactly, the behavior you referenced via `g/` from the previous PR is the one I tried to describe. Hopefully it is possible to reuse the code from this PR and implement both types of jumps, as these two seem to be the most popular approaches for blazing fast movement.

**Comment by postsolar** [comment]
Re: how it acts with regards to selections. An interesting implementation has been recently tried out in https://github.com/phaazon/hop.kak. What I found really elegant and interesting is that it's extensible as in it only acts as a "selections filter" so by using selections registers the user is free to build their own command, be it one adding a new cursor, or extending current selection, or something else

**Comment by the-mikedavis** [comment]
Ah right, I had forgotten that the number of characters to search is configurable in #5340. That reminds me why I'm not enthusiastic about adding character based jumping or jump features in general to core.

What I don't like about #5340 is the size of the feature: it's adds a lot of code (four new modules, state-machine-like code, cosmetics like dimming), a new configuration section under `editor.jump-mode`, and basically adds a new "mode" to the editor (where this PR uses established machinery: just `cx.on_next_key`). I'm hesitant to add jumping to core at all because there are so many methods of doing it: there are many existing plugins and each plugin has a opinions on how to jump. I'm imagining us ending up with a lot of code in core covering multiple methods of jumping and/or behavior configuration like number of characters because we're trying to satisfy all the ways people think a jump feature should work.

So if we add jumping at all I'd like to keep it very simple like this Amp-like jump and leave everything else to plugins.

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5908]: When jumping in normal mode, should we normalize the direction of the selection like we do for LSP jumping (like we discussed in https://github.com/helix-editor/helix/discussions/6500#discussioncomment-5492299)?

Currently if you're jumping ahead the direction is forward and behind is backwards which makes sense for extending but I think we could always use backwards for normal mode jumping.

Comment (pascalkuthe) [reply helix-term/src/commands.rs]: yeah good point

File: helix-view/src/editor.rs
Comment (pascalkuthe) [reply helix-view/src/editor.rs:317]: done, I didn't change the default because I felt there was no good default choice for qwerty its subjective defending on handsize and so forth.

**Comment by yerlaser** [comment]
Tested it and it works beautifully.
Do you, guys, have a document/wiki that describes what a proper reviewer should do and how to become one?

**Comment by the-mikedavis** [comment]
You don't need any permissions to review PRs, feel free to leave review comments on open PRs. We don't have an exact procedure for becoming a maintainer / gaining triage or write priviledges if that's what you mean. This PR isn't really waiting on review though: the changes look good but it'd be nice to get some input from @semin-park (see https://github.com/helix-editor/helix/pull/8875#pullrequestreview-1743457571). He's been inactive for quite some time though so I think it would be ok to go ahead with this (\cc @archseer)

**Comment by dvic** [comment]
For what it's worth: I've been using this for a few months and have not encountered any issues with it.

**Comment by wontem** [comment]
One minor bug I encountered immediately: the last line in the buffer has an invalid number of anchors unless it's new-line-terminated:

rust, `Ok` is not reachable if there's no new line (not formatted correctly just for the demo purpose):





toml, `"all"` is not reachable if there's no new line:




```
[editor.whitespace]
render = "all"
```

**Comment by pascalkuthe** [comment]
I can't reproduce that

**Comment by wontem** [comment]
I apologize, my fault, pulled from the `helix:amp-jump` instead of `pascalkuthe:amp-jump`. It works like a charm, I will never leave this branch now 😄

**Comment by wontem** [comment]
Now I'm on the correct version. I wondered what if I mistyped a command, so I typed `gw` and then `bw` (which is not present across the variants). I expected I stay in the same place, but I jumped into a random position. Then I repeated and I jumped somewhere else again. I checked with other absent combinations (`az, aw etc`) - the result is the same

**Comment by the-mikedavis** [comment]
I can't reproduce that. If you hit a key that can't be a jump label then we bail out:

https://github.com/pascalkuthe/helix/blob/48d2638c031b3e17f60fbf82d1388584d43fc31d/helix-term/src/commands.rs#L5656-L5660

https://github.com/pascalkuthe/helix/blob/48d2638c031b3e17f60fbf82d1388584d43fc31d/helix-term/src/commands.rs#L5664-L5666

Can you make an https://asciinema.org/ recording / video / gif of what you see?

**Comment by wontem** [comment]
I have no `asciinema` configured, going to do it a bit later. For now, hopefully, this video will suffice:

https://github.com/helix-editor/helix/assets/7056524/7220d08c-e9ea-4c84-8da5-9c557b289995

toml
```toml
theme = "custom"

[editor]
line-number = "relative"
mouse = false
bufferline = "always"
true-color = true
auto-format = true
auto-save = true
```

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5725]: Ah looks like we have some duplicates in the ranges:

<details><summary>Ranges from the reproduction case...</summary>

```rust
[
	Range { anchor: 19, head: 25, .. },
	Range { anchor: 15, head: 9, .. },
	Range { anchor: 19, head: 25, .. },
	Range { anchor: 27, head: 31, .. },
	Range { anchor: 27, head: 31, .. },
	Range { anchor: 5, head: 0, .. },
	Range { anchor: 32, head: 38, .. },
	Range { anchor: 32, head: 39, .. },
	Range { anchor: 42, head: 50, .. },
	Range { anchor: 42, head: 50, .. },
	Range { anchor: 52, head: 57, .. },
	Range { anchor: 52, head: 58, .. },
	Range { anchor: 60, head: 65, .. },
	Range { anchor: 66, head: 76, .. },
	Range { anchor: 66, head: 77, .. },
	Range { anchor: 80, head: 86, .. },
	Range { anchor: 80, head: 86, .. },
	Range { anchor: 88, head: 92, .. },
	Range { anchor: 88, head: 92, .. },
	Range { anchor: 93, head: 98, .. },
	Range { anchor: 93, head: 99, .. },
	Range { anchor: 101, head: 105, .. },
	Range { anchor: 106, head: 110, .. },
	Range { anchor: 106, head: 110, .. },
	Range { anchor: 111, head: 117, .. },
	Range { anchor: 111, head: 118, .. },
	Range { anchor: 120, head: 124, .. },
	Range { anchor: 125, head: 129, .. },
	Range { anchor: 125, head: 129, .. },
	Range { anchor: 130, head: 134, .. },
	Range { anchor: 130, head: 135, .. },
	Range { anchor: 137, head: 141, .. },
	Range { anchor: 137, head: 141, .. }
]
```

</details>

`gwaw` there is going to a valid label, it's just not visible since that one is overlapped by another label.

It looks like the `eb`-like movement trips over punctuation and can re-select the last word

**Comment by wontem** [comment]
Another quick question, is it necessary to have an anchor in the current cursor position (if you're already in the beginning of a word)?

File: runtime/themes/everforest_dark.toml
Comment (James-Firth) [inline runtime/themes/everforest_dark.toml:1]: I'm not sure what's common for the scope of theme changes like this but is it worth adding to more themes during this PR if it's going to be accepted?

Looking at Solarized Dark
image

and Dracula (my personal favourite)
image

The text doesn't end up changing. That said not all of them stay the same, github dark changes (though hard to read)
image
or Adwaita looks fine
image

**Review by James-Firth — COMMENTED** [review]
Docs are clear and the feature works great! (Just did my first build to test this out!)

Just have one comment/question about themes

File: runtime/themes/everforest_dark.toml
Comment (david-crespo) [reply runtime/themes/everforest_dark.toml:1]: While most people will probably want the color to change, it's still usable without that, and I think it's the most reasonable default when no color is configured because there's no way to generically choose the right color for all themes. Not sure why some would change without a color configured, though.

Comment (James-Firth) [reply runtime/themes/everforest_dark.toml:1]: @david-crespo I just did some poking around, adding `"ui.virtual" = "colourname"` adds the functionality since the jump mode is using [virtual text, noted here](https://github.com/helix-editor/helix/pull/8875/files#diff-d782c07f22492bcae17333aba5dc5b840fb30317727e71267ebf4863f84787fdR5625).

Example adding this line to Solarized Dark and then activating jump mode
image
image

Seeing this, I actually no longer have any comment about the colours. My original point was wondering if more themes should be updated in advance for broader support, but really this just highlighted to me which themes don't have colour support for virtual text.

File: helix-term/src/commands.rs
Comment (pascalkuthe) [reply helix-term/src/commands.rs]: It took some time to dig trough this but I fixed this and cleaned up some edgecases

File: runtime/themes/everforest_dark.toml
Comment (pascalkuthe) [reply runtime/themes/everforest_dark.toml:1]: yeah we don't require new prs to edit themes. I changed the default theme other themes can be adopted in separate prs once this is merged

**Comment by postsolar** [comment]
I noticed that, unlike `search_next` and similar commands, `goto_word` does not respect cursor orientation.

**Comment by pascalkuthe** [comment]
It does in visual Mode just not on normal mode. It's intentional.

**Comment by postsolar** [comment]
Got it, but what is the reasoning for this?

**Comment by the-mikedavis** [comment]
See my comment above: https://github.com/helix-editor/helix/pull/8875#discussion_r1428152259

It matches the LSP location jumping features

File: helix-core/src/graphemes.rs
Comment (the-mikedavis) [inline helix-core/src/graphemes.rs:428]:
```suggestion
/// An iterator over the graphemes of a `RopeSlice` in reverse.
```

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:5891]: Looks like maybe a stray comment here

**Review by the-mikedavis — APPROVED** [review]

**Review by archseer — APPROVED** [review]

**Comment by lucmos** [comment]
Thank you for this feature! It's working great :smile: 

Just a quick question -- Is it possible to "add" a selection to the word I am jumping to? Without extending the current selection as in visual mode, i.e., an easy way to add multiple cursors.

**Comment by the-mikedavis** [comment]
There's some discussion about that above: https://github.com/helix-editor/helix/pull/8875#issuecomment-1845500723

**Comment by pblkt** [comment]
It's likely that your theme doesn't define the [`ui.virtual.jump-label`](https://docs.helix-editor.com/themes.html#scopes:~:text=ui.virtual.jump,virtual%20jump%20labels) or the fallback `ui.virtual`. 

To verify that this is indeed the issue, check that the labels in `:theme default` are highlighted as expected .

To fix, make sure your theme defines it by either creating and using a local one or opening a PR.

**Comment by giann** [comment]
That was it thank you

**Comment by markstos** [comment]
> danillos I have a question, would it be possible to work in all visible buffers? So we could go to a word in another split window?

> @the-mikedavis  jumping across views is not necessary to optimize for - I think switching views and then jumping within the view would work well enough.

Allowing this to work across all visible windows could be an elegant way to directly switch to another window to reach ones that  can't be directly access by simply going "left/right/up/down" or to the last window.

wmfocus, sway-easyfocus, tmux, and some vim plugins all offer a way to switch to another window by selecting a letter. Here's the concept implemented in wmfocus (if the gif animates):

image

This allows navigating directly to any visible window-- helpful on 4k screens where 6 or Helix windows can fit in a terminal. 

By extending it to every visible window, the feature would be useful for both "amp-like jumping to a location in the current window" as well "switching to another window by label".

**Comment by pickfire** [comment]
I noticed there is a bug where jump label switches positions with the text when cjk characters is involved, which causes layout shift, making it hard to know where to jump to.

I think it should preserves the existing positions if cjk characters is involved, otherwise the user need to read the whole thing to know where to jump to again.

For example

截屏2024-07-19 上午9 38 56

After `gw`, it becomes

截屏2024-07-19 上午9 39 13

In the second line, it should pad the first cjk character so it doesn't do layout shift either. The same goes for emoji and double width characters.

**Pull request merged**
